core.ts

```diff
@@ -0,0 +1,1039 @@
+const noop = () => {}
+const deprecated: {
+	[ index: string ]: () => void
+} = {}
+const addDeprecated = (name: string, message: string) => {
+	deprecated[name] = () => {
+		console.warn('pinejs-client deprecated:', message)
+		deprecated[name] = noop
+	}
+}
+addDeprecated(
+	'expandFilter',
+	'`$filter: a: b: ...` is deprecated, please use `$filter: a: $any: { $alias: "x", $expr: x: b: ... }` instead.'
+)
+
+function defaults <T>(a: T | undefined, b: T | undefined, z: T): T
+function defaults <T>(a: T | undefined, z: T): T
+function defaults <T>(...args: (T | undefined)[]): T | undefined {
+	for (const arg of args) {
+		if (arg != null) {
+			return arg
+		}
+	}
+}
+const map = <T, R>(arr: T[], fn: (value: T) => R): R[] => {
+	const results = []
+	for (const key of arr) {
+		results.push(fn(key))
+	}
+	return results
+}
+const mapObj = <T, R>(obj: {[index: string]: T}, fn: (value: T, key: string) => R): R[] => {
+	const results = []
+	for (const key in obj) {
+		if (obj.hasOwnProperty(key)) {
+			results.push(fn(obj[key], key))
+		}
+	}
+	return results
+}
+
+const requiredUtilMethods = ['isString', 'isNumber', 'isBoolean', 'isObject', 'isArray', 'isDate']
+const isUtil = (obj: any): obj is PinejsClientCoreFactory.Util => {
+	if (obj == null) {
+		return false
+	}
+	for (const method of requiredUtilMethods) {
+		if (obj[method] == null) {
+			return false
+		}
+	}
+	return true
+}
+const isPromiseRejector = (obj: any): obj is PinejsClientCoreFactory.PromiseRejector => {
+	return obj != null && obj.reject != null
+}
+const isValidOption = (key: string): key is keyof PinejsClientCoreFactory.ODataOptions => {
+	return key === '$filter' ||
+		key === '$expand' ||
+		key === '$orderby' ||
+		key === '$top' ||
+		key === '$skip' ||
+		key === '$select'
+}
+
+// Workaround the fact that `setInterval` returns a different type in nodejs vs browsers
+// TODO: typescript 2.8 will introduce `ReturnType` as a better way to do this
+const _setIntervalResult = false as true && setInterval(() => {}, 0)
+
+type PollOnObj = {
+	unsubscribe: () => void
+}
+class Poll<PromiseResult extends PromiseLike<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]> = Promise<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]>> {
+	private subscribers: {
+		error: Array<(response: PromiseResult) => void>,
+		data: Array<(err: any) => void>,
+	} = {
+		error: [],
+		data: [],
+	}
+
+	private stopped = false
+	private pollInterval?: typeof _setIntervalResult
+
+	private requestFn: null | (() => PromiseResult)
+
+	constructor(
+		requestFn: () => PromiseResult,
```

Comment:
Any reason not to specify that one as `private requestFn: ...` as well?

Reply:
Because it needs to be able to be null, since it gets set to `null` in `destroy` but to type it as nullable in the constructor would allow null to be passed in which would be an error